WhateverGreen/kern_igfx.cpp

```diff
@@ -465,53 +452,102 @@ bool IGFX::wrapAcceleratorStart(IOService *that, IOService *provider) {
 }
 
-IOReturn IGFX::wrapHwSetBacklight(void *that, uint32_t backlight) {
-	// AppleIntelFramebufferController::hwSetBacklight
-	// Use our backlight calculation here and skip call to original function to avoid flicker
+void IGFX::wrapCflWriteRegister32(void *that, uint32_t reg, uint32_t value) {
+	if (reg == BXT_BLC_PWM_FREQ1) {
+		if (value && value != callbackIGFX->driverBacklightFrequency) {
+			DBGLOG("igfx", "wrapCflWriteRegister32: driver requested BXT_BLC_PWM_FREQ1 = 0x%x", value);
+			callbackIGFX->driverBacklightFrequency = value;
+		}
 
-	callbackIGFX->backlightLevel = backlight;
+		if (callbackIGFX->targetBacklightFrequency == 0) {
+			// Save the hardware PWM frequency as initially set up by the system firmware.
+			// We'll need this to restore later after system sleep.
+			callbackIGFX->targetBacklightFrequency = callbackIGFX->orgCflReadRegister32(that, BXT_BLC_PWM_FREQ1);
+			DBGLOG("igfx", "wrapCflWriteRegister32: system initialized BXT_BLC_PWM_FREQ1 = 0x%x", callbackIGFX->targetBacklightFrequency);
 
-#ifdef DEBUG
-	uint64_t bxt_blc_pwm_ctl1 = callbackIGFX->orgReadRegister32(that, BXT_BLC_PWM_CTL1);
-#endif
-	
-	// Calculate backlight duty in 64-bits so result is not truncated (essentially CFL backlight fix)
-	uint64_t bxt_blc_pwm_freq1 = callbackIGFX->backlightFrequency;
-	uint64_t bxt_blc_pwm_duty1 = static_cast<uint64_t>(callbackIGFX->backlightLevel) * bxt_blc_pwm_freq1 / 0xFFFFLL;
-	
-	callbackIGFX->orgWriteRegister32(that, BXT_BLC_PWM_FREQ1, static_cast<uint32_t>(bxt_blc_pwm_freq1));
-	callbackIGFX->orgWriteRegister32(that, BXT_BLC_PWM_DUTY1, static_cast<uint32_t>(bxt_blc_pwm_duty1));
-	
-	DBGLOG("igfx", "wrapHwSetBacklight(): BXT_BLC_PWM_CTL1=0x%llX BXT_BLC_PWM_FREQ1=0x%llX BXT_BLC_PWM_DUTY1=0x%llX backlightLevel=0x%X",
-		   bxt_blc_pwm_ctl1, bxt_blc_pwm_freq1, bxt_blc_pwm_duty1, callbackIGFX->backlightLevel);
-	
-	return kIOReturnSuccess;
+			if (callbackIGFX->targetBacklightFrequency == 0) {
+				// This should not happen with correctly written bootloader code, but in case it does, let's use a failsafe default value.
+				callbackIGFX->targetBacklightFrequency = FallbackTargetBacklightFrequency;
+				SYSLOG("igfx", "wrapCflWriteRegister32: system initialized BXT_BLC_PWM_FREQ1 is ZERO");
+			}
+		}
+
+		if (value) {
+			// Nonzero writes to this register need to use the original system value.
+			// Yet the driver can safely write zero to this register as part of system sleep.
+			value = callbackIGFX->targetBacklightFrequency;
+		}
+	} else if (reg == BXT_BLC_PWM_DUTY1) {
+		if (callbackIGFX->driverBacklightFrequency && callbackIGFX->targetBacklightFrequency) {
+			// Translate the PWM duty cycle between the driver scale value and the HW scale value
+			uint32_t rescaledValue = static_cast<uint32_t>(value /
+														   (static_cast<uint64_t>(callbackIGFX->driverBacklightFrequency) *
+														   static_cast<uint64_t>(callbackIGFX->targetBacklightFrequency)));
```

Comment:
I think you made a transcription error here or something, this math is going to always rescale the value to zero. If you don't want to use the floating-point math in my original patch, this integer version should work:
```
 uint32_t rescaledValue = (value * static_cast<uint64_t>(callbackIGFX->targetBacklightFrequency)) / static_cast<uint64_t>(callbackIGFX->driverBacklightFrequency);
```